src/printer.js

```diff
@@ -1614,15 +1614,18 @@ function genericPrintNoParens(path, options, print) {
         return join(" & ", types);
       }
 
-      return group(
-        indent(
-          options.tabWidth,
-          concat([
-            ifBreak(concat([line, isIntersection ? "&" : "|", " "])),
-            join(concat([line, isIntersection ? "&" : "|", " "]), types)
-          ])
-        )
-      );
+      const parent = path.getParentNode();
+      // If there's a leading comment, the parent is doing the indentation
+      const shouldIndent = !(parent.type === 'TypeAlias' &&
+        hasLeadingOwnLineComment(options.originalText, n));
+
+      const token = isIntersection ? "&" : "|";
+      const printed = concat([
+        ifBreak(concat([shouldIndent ? line : "", token, " "])),
+        join(concat([line, token, " "]), types)
+      ]);
+
+      return group(shouldIndent ? indent(options.tabWidth, printed) : printed);
```

Comment:
I was talking about something like this. It's nice to keep is all a single expression as I think it's more clear how it's laid out:

```js
return group(
  indent(
    shouldIndent ? options.tabWidth : 0,
    concat([
      ifBreak(concat([line, isIntersection ? "&" : "|", " "])),
      join(concat([line, isIntersection ? "&" : "|", " "]), types)
    ])
  )
);
```

I copied the code from before your PR so if you've tweaked this it may not have your updates, but I don't see why this shouldn't work.